### PR TITLE
Implemented Feature to download all Images by Public User on Insta. Fixed #16

### LINKED ISSUE #16: Add a Feature to Download All the Post of A Public Instagram Profile.
# Add a Feature to Download All the Post of A Public Instagram Profile.

## Overview
Add a Feature to Download All the Post of A Public Instagram Profile and Send it to the User.

## Description
The Goal is to Create a Feature in pastebin-telegram bot to download all the Post specifically images to the server and send it to the user afterwards, Use threading to Download Photos concurrently at the same time.

## Expected outcomes

Feature of downloading all Post images of a specific user from Instagram using Pastebin-telegram.
## Skills

Python, Threading, Api's.

## Possible mentors
[Kanishk Pachauri](https://github.com/Mr-Sunglasses), [Anjuman Hassan](https://github.com/AnjumanHasan)

## Expected project size

100 hours.

## Difficulty

Hard.

### PR DESCRIPTION
**Pull request by devAyushDubey**
Hey @Mr-Sunglasses , I tried a few things, and it's working perfectly on my test bot.
Used threading to multitask fetching of images from Instagram.
fetch_posts function in [insta_all_pictures.py ](https://github.com/devAyushDubey/pastebin-telegram/blob/master/pastebin-bot/insta_all_pictures.py) uses multiple threads to download pictures from Instagram.
Invalid user name exception is also handled and a message is sent in response in the [bot.py](https://github.com/devAyushDubey/pastebin-telegram/blob/master/pastebin-bot/bot.py).

### CODE REVIEW

**Comment by Mr-Sunglasses** [comment]
@devAyushDubey Do let me know if it is working without login using instagram.

File: pastebin-bot/insta_all_pictures.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import threading
+import shutil
+
+profile_instance = Instaloader()
+
+
+def fetch_posts(user):
+
+    try:
+        prof = Profile.from_username(profile_instance.context, user)
```
Comment (Mr-Sunglasses) [inline pastebin-bot/insta_all_pictures.py:23]: As we are already using a meta, account for handling the username and password of insta to efficient use of Instaloader module so try to use `profile_instance.login(user=Username,passwd=Passoword)`

**Comment by devAyushDubey** [comment]
> @devAyushDubey Do let me know if it is working without login using instagram.

Yes @Mr-Sunglasses this does not require any login.

**Comment by Mr-Sunglasses** [comment]
@devAyushDubey I'm trying it to run locally and it is not working on my system, can you please share any demo that it is working in your system.

**Comment by devAyushDubey** [comment]
> @devAyushDubey I'm trying it to run locally and it is not working on my system, can you please share any demo that it is working in your system.

You can test it on https://t.me/AyushOS_bot between 11 to 12:30, that's when I will get access to my laptop. I will run it locally so that you can test it. If you want API KEY to this bot for testing do let me know I can share that with you on anywhere private (Whatsapp /Discord).

Are you getting some sort of error btw?

**Comment by Mr-Sunglasses** [comment]
> > @devAyushDubey I'm trying it to run locally and it is not working on my system, can you please share any demo that it is working in your system.
> 
> You can test it on https://t.me/AyushOS_bot between 11 to 12:30, that's when I will get access to my laptop. I will run it locally so that you can test it. If you want API KEY to this bot for testing do let me know I can share that with you on anywhere private (Whatsapp /Discord).
> 
> Are you getting some sort of error btw?

So the problem is The bot runs on the virtual machine in the server and when it request to Instagram and scrape the data, Instagram blocks it due to Proxy and IP address issues, that is why I'm asking you to use ID and password for scrapping. And also my computer exhausted all the requests limits 😅of Instaloader so it is not working on my computer.

**Comment by Mr-Sunglasses** [comment]
image
@devAyushDubey  I think this is programmed differently 😂 😂

**Comment by devAyushDubey** [comment]
> image @devAyushDubey I think this is programmed differently 😂 😂

I think we should ignore that @Mr-Sunglasses 😂. Just hostel things.

**Comment by devAyushDubey** [comment]
I have added login in the last commit. @Mr-Sunglasses , do let me know if it works. And sorry for that greeting message, was not meant for you 😂, I hope it's chill.

**Comment by Mr-Sunglasses** [comment]
image
@devAyushDubey 
It is just not working on my device 🙃.

**Comment by devAyushDubey** [comment]
Yeah @Mr-Sunglasses  I was doing some changes that time. You should checkout the [test bot](https://t.me/AyushOS_bot) now. I hosted it on a server and made some changes. Of course added Login, and Rate Limit, I am limiting it to usernames with <= 200 posts.
Instaloader and Instagram API behaves weird sometimes, there were many issues for the same on Instaloader's repo too. While the bot would work fine, there could be 400 and 401 sometimes randomly. I have added different delays before every request to minimize this. Try it on the test bot and do let me know so that I can make a commit.

And yeah I changed that greeting message 🤦‍♀️ to Hi User! 😂

**Comment by devAyushDubey** [comment]
WhatsApp Image 2022-09-19 at 10 53 35 PM

**Review by Mr-Sunglasses — APPROVED** [review]
Looking Good we can now merge it.

**Review by AnjumanHasan — APPROVED** [review]
LGTM

**Comment by Mr-Sunglasses** [comment]
Thanks, @devAyushDubey, I Will Looking forward to Improving this feature in the next version.

**Comment by devAyushDubey** [comment]
Screenshot (950)
@Mr-Sunglasses Just got my account blocked (not main one) 😂😂. I had that last bit of code that added that rate limit and delays, it also optimized the login but I don't think I would be able to test it now 😂, so let that be as it is for now.

Was nice contributing to pastebot 😉.

**Pull request merged**